junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala

```diff
@@ -0,0 +1,236 @@
+          s"$msg, took $TIME_TAG sec")
+    }
+
+    def testAssumptionViolatedOutput(method: String): Output = {
+      Warn(s"Test assumption in test $formattedTestClass.${red(method)}" +
+          s" failed: $formattedAssumptionViolatedException:" +
+          s" This assume should not pass, took $TIME_TAG sec")
+    }
+
+    def testFinishedOutput(method: String): Output = {
+      Debug(s"Test $formattedTestClass.${cyan(method)} finished, " +
+          s"took $TIME_TAG sec")
+    }
+
+    def testRunStartedOutput: Output =
+      infoIfOrElseDebug(verbose, blue("Test run started"))
+
+    def testRunFinishedOutput: Output = {
+      val buff = new StringBuilder
+      buff.append(blue("Test run finished: "))
+      val failedColor = if (expectedFail == 0) blue _ else red _
+      buff.append(failedColor(s"$expectedFail failed"))
+      buff.append(blue(", "))
+      val ignoredColor = if (expectedIgnored == 0) blue _ else yellow _
+      buff.append(ignoredColor(s"$expectedIgnored ignored"))
+      buff.append(blue(s", $expectedTotal total, ${TIME_TAG}s"))
+      infoIfOrElseDebug(verbose, buff.result())
+    }
+
+    def skippedEvent: Output = Event("Skipped")
+    def successEvent: Output = Event("Success")
+    def failureEvent: Output = Event("Failure")
+
+    def infoIfOrElseDebug(p: Boolean, msg: String): Output =
+      if (p) Info(msg)
+      else Debug(msg)
+
+    override def toString: String = frameworkArgs.mkString("[", ", ", "]")
+
+    private def red(str: String): String =
+      if (noColor) str
+      else RED + str + NORMAL
+
+    private def yellow(str: String): String =
+      if (noColor) str
+      else YELLOW + str + NORMAL
+
+    private def blue(str: String): String =
+      if (noColor) str
+      else BLUE + str + NORMAL
+
+    private def magenta(str: String): String =
+      if (noColor) str
+      else MAGENTA + str + NORMAL
+
+    private def cyan(str: String): String =
+      if (noColor) str
+      else CYAN + str + NORMAL
+
+  }
+
+  protected class JunitTestRecorder(context: OutputContext)
+      extends Logger with EventHandler {
+    val buff = List.newBuilder[Output]
+
+    def ansiCodesSupported(): Boolean = true
+
+    def info(msg: String): Unit = buff += Info(encodeTimes(msg))
+
+    def warn(msg: String): Unit = buff += Warn(encodeTimes(msg))
+
+    def debug(msg: String): Unit = buff += Debug(encodeTimes(msg))
+
+    def error(msg: String): Unit = {
+      if (msg.startsWith("    at ") || msg.startsWith("    ...")) {
+        println(GREY + "Ignoring stacktrace output: " + msg + NORMAL)
+      } else {
+        buff += Error(encodeTimes(msg))
+      }
+    }
+
+    def trace(t: Throwable): Unit = buff += Trace(t.getMessage)
+
+    def handle(event: sbt.testing.Event): Unit = {
+      buff += Event(event.status().toString)
+    }
+
+    def recordDone(msg: String): Unit = {
+      buff += Done(msg)
+      checkOutput(buff.result())
```

Comment:
IMO `checkOutput` should be called explicitly from `testJUnitOutput`, rather than implicitly through `recordDone`.